### PR TITLE
feat(deploy): deploy auto-populates SP, SSH, DNSPrefix

### LINKED ISSUE #872: Polish `deploy` command
Some cleanup + polish tasks for the `deploy` command before we document and promote it more:
 
- [ ] Support auto-fill for SSH key
- [ ] Support auto-fill for cluster name (derived from RG name)
- [ ] Support auto-fill for SP creds
- [ ] Cleanup UX for the output (there’s a few places where the wrong output formatter is used and looks sloppy)

### PR DESCRIPTION
**Pull request by colemickens**
**What this PR does / why we need it**: Implements some automatic default setting for our example apimodels. The result is that, at least for the basic example files, they can be deployed immediately without the user needing to provision credentials or edit the apimodel files.

**Which issue this PR fixes**: fixes #872

**Special notes for your reviewer**: 
**Release note**:

```release-note
implement automatic deploy of simple example apimodels
```

This is how I'm testing this change (with no changes to the `kubernetes.json` - it has an empty dnsprefix, empty ssh key and empty serviceprincipal profile)

```
make
./acs-engine deploy \
  --dns-prefix=colemick \
  --auto-suffix \
  --location=westus2 \
  --subscription-id=c1089427-83d3-4286-9f35-5af546a6eb67 \
    ./examples/kubernetes.json
```

cc: @seanknox

---
This change is [Reviewable](https://reviewable.io/reviews/azure/acs-engine/925)

### CODE REVIEW

**Comment by colemickens** [comment]
Example output:

```
WARN[0000] apimodel: missing masterProfile.dnsPrefix will use "colemick-595eccd1" 
WARN[0000] --resource-group was not specified. Using the DNS prefix from the apimodel as the resource group name: colemick-595eccd1 
WARN[0005] apimodel: ServicePrincipalProfile was empty, creating application... 
WARN[0006] apimodel: ServicePrincipalProfile was empty, assigning role to application... 
WARN[0007] Failed to create role assignment (will retry): "authorization.RoleAssignmentsClient#Create: Failure responding to request: StatusCode=400 -- Original Error: autorest/azure: Service returned an error. Status=400 Code=\"PrincipalNotFound\" Message=\"Principal 8d0fdd5227264bd69edecb7207e08057 does not exist in the directory 72f988bf-86f1-41af-91ab-2d7cd011db47.\"" 
WARN[0010] Failed to create role assignment (will retry): "authorization.RoleAssignmentsClient#Create: Failure responding to request: StatusCode=400 -- Original Error: autorest/azure: Service returned an error. Status=400 Code=\"PrincipalNotFound\" Message=\"Principal 8d0fdd5227264bd69edecb7207e08057 does not exist in the directory 72f988bf-86f1-41af-91ab-2d7cd011db47.\"" 
INFO[0026] Starting ARM Deployment (colemick-595eccd1-839951915). This will take some time... 
INFO[0387] Finished ARM Deployment (colemick-595eccd1-839951915). 
```

And the output directory contains a new `id_rsa` for the cluster.

I'll patch up the test later tonight.

**Comment by colemickens** [comment]
It passes at least, but I plan to rechunk it so it's easier to review, there's sort of just one massive blob right now.

File: cmd/deploy.go [hunk 1]
```diff
@@ -27,6 +28,8 @@ type deployCmd struct {
 	authArgs
 
 	apimodelPath      string
+	dnsPrefix         string
+	autoSuffix        bool
```
Comment (seanknox) [inline cmd/deploy.go]: Do we have any unit tests for `deploy.go`?

File: pkg/acsengine/ssh.go [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+	"golang.org/x/crypto/ssh"
+)
+
+const (
+	// SshKeySize is the size of SSH key to create
+	SSHKeySize = 4096
+)
+
+// CreateSaveSSH generates and stashes an SSH key pair.
+func CreateSaveSSH(username, outputDirectory string) (privateKey *rsa.PrivateKey, publicKeyString string, err error) {
```
Comment (seanknox) [inline pkg/acsengine/ssh.go]: Possibly to get some tests around this too, even something as simple checking method signature + output?

**Comment by colemickens** [comment]
I can add some unit tests around SSH at least. I'll take a look at `deploy.go` too, but I think it's a pretty shallow wrapper around the Go SDK DeploymentsClient.

Thank you for taking a look @seanknox! I'm still planning to chunk this and add some unit tests as soon as I can make time. Thx.

File: cmd/deploy.go (hunk 1 above)
Comment (colemickens) [inline cmd/deploy.go]: Not right now, I think integration test might be more important. In terms of what I'd unit test in deploy.go, the issue I filed about changing validation would let us formalize how the default setting happens in a much more testable fashion.

**Comment by colemickens** [comment]
@sean SSH test added. Other question answered. Rebased so commits are reviewable/smaller. Should be good to go once Jenkins turns green. (If you're cool with it, can you Approve?)

**Comment by seanknox** [comment]
@colemickens merging this in now so you don't have to wait when you're back.

**Pull request merged**